bdk_cbf_example/src/main.rs

```diff
@@ -0,0 +1,158 @@
+        bdk_cli::Commands::ChainSpecific(cbf_cmd) => cbf_cmd,
+        general_cmd => {
+            return bdk_cli::handle_commands(
+                general_cmd,
+                client,
+                &mut keychain_tracker,
+                &mut db,
+                args.network,
+                &keymap,
+            );
+        }
+    };
+
+    match cbf_cmd {
+        CbfCommands::Reveal { size } => {
+            let mut keychain_tracker = keychain_tracker.lock().unwrap();
+            let targets = keychain_tracker
+                .keychains()
+                .into_iter()
+                .map(|(k, _)| (*k, size))
+                .collect();
+            let (_, derivation_additions) = keychain_tracker
+                .txout_index
+                .reveal_to_target_multi(&targets);
+            let changeset = KeychainChangeSet {
+                derivation_indices: derivation_additions,
+                ..Default::default()
+            };
+            db.lock().unwrap().append_changeset(&changeset)?;
+            keychain_tracker.apply_changeset(changeset);
+            Ok(())
+        }
+        CbfCommands::Rescan { _batch_size } => {
+            todo!("Implement rescan from sync");
+            // This function will reveal batch_size scripts and rescan the nakamoto
+            // client
+            // If we notice, after rescan, that more than `batch_size - epsilon` scripts
+            // have been used, we rescan using 2*batch_size, and so on
+        }
+        CbfCommands::Scan => {
+            // indexing logic!
+            let mut keychain_tracker = keychain_tracker.lock().unwrap();
+
+            // find scripts!
+            let scripts = keychain_tracker
+                .txout_index
+                .revealed_spks_of_all_keychains()
+                .into_values()
+                .flatten()
+                .map(|(_, spk)| spk.clone());
+
+            let last_checkpoint = keychain_tracker.chain_graph().chain().latest_checkpoint();
+
+            let mut update = ChainGraph::<TxHeight>::default();
+
+            if let Some(last_cp) = last_checkpoint.clone() {
+                // Otherwise the chains don't connect :)
+                let _ = update.insert_checkpoint(last_cp.clone())?;
+            }
+
+            let mut derivation_additions = DerivationAdditions::default();
+
+            client.sync_setup(scripts, last_checkpoint)?;
+            loop {
+                match client.next_event()? {
+                    CbfEvent::BlockMatched(id, txs) => {
+                        let _ = update.insert_checkpoint(BlockId {
+                            height: id.height as u32,
+                            hash: id.hash,
+                        })?;
+
+                        for (tx, height) in txs {
+                            //keychain_tracker.lock().unwrap().txout_index.pad_all_with_unused(stop_gap);
+                            if keychain_tracker.txout_index.is_relevant(&tx) {
```

Comment:
I think rather than doing this here we should simply have a way of filtering out irrelevant things out of an update. For now we can just include all transactions in the block. I made an issue to build the tools to do this properly: https://github.com/LLFourn/bdk_core_staging/issues/193

There really isn't much harm to just acquiring a `.lock()` on the db and the keychain tracker and just doing an `apply_update` to the `KeychainTracker` and pushing the thing into the database right away. This is good because it saves your progress for each block as you go.